pkg/executor/insert.go

```diff
@@ -301,15 +354,51 @@ func (e *InsertExec) batchUpdateDupRows(ctx context.Context, newRows [][]types.D
 				return err
 			}
 
-			newRows[i] = nil
-			break
+			shouldRemoveOldRow := false
+			if len(e.replaceConflictIfExpr) > 0 {
+				shouldRemoveOldRow, err = e.replaceConflictIf(e.Ctx().GetExprCtx().GetEvalCtx(), oldRow)
+				if err != nil {
+					return err
+				}
+			}
+
+			if shouldRemoveOldRow {
+				removeOldRows = append(removeOldRows, removeOldRow{
+					handle: handle,
+					oldRow: oldRow,
+				})
+			} else if len(e.OnDuplicate) > 0 {
+				if err = e.updateDupRow(ctx, i, r, oldRow, handle, updateDupKeyCheck, autoColIdx); err != nil {
+					return err
+				}
+				newRows[i] = nil
+				break
+			} else if e.ignoreErr {
+				e.Ctx().GetSessionVars().StmtCtx.AppendWarning(uk.dupErr)
+				if txnCtx := e.Ctx().GetSessionVars().TxnCtx; txnCtx.IsPessimistic &&
+					e.Ctx().GetSessionVars().LockUnchangedKeys {
+					// lock duplicated row key on insert-ignore
+					txnCtx.AddUnchangedKeyForLock(uk.newKey)
+				}
+				newRows[i] = nil
+				break
+			}
 		}
 
 		// If row was checked with no duplicate keys,
 		// we should do insert the row,
 		// and key-values should be filled back to dupOldRowValues for the further row check,
 		// due to there may be duplicate keys inside the insert statement.
 		if newRows[i] != nil {
+			for _, rm := range removeOldRows {
```

Comment:
Could you add some comments that why the `removeOldRows` should not be used in ignore or update mode?

- In ignore mode, the inserted row is not applied, so we should not delete the related tombstone rows.
- In update mode, the tombstone conflict rows may not be the same as those in insert mode; we should recompute them.